src/geometry/manifold/manifold.lean

```diff
@@ -0,0 +1,539 @@
+    simp only [exists_prop, mem_Union, mem_singleton_iff] at ht,
+    rcases ht with ⟨e', e'_atlas, s, s_open, ts⟩,
+    rw ts,
+    let f := e.symm.trans e',
+    have : is_open (f.to_fun ⁻¹' s ∩ f.source),
+      by simpa [inter_comm] using (continuous_on_open_iff (c.open_source e e' he e'_atlas)).1
+        (c.continuous_to_fun e e' he e'_atlas) s s_open,
+    have A : e'.to_fun ∘ e.inv_fun ⁻¹' s ∩ (e.target ∩ e.inv_fun ⁻¹' e'.source) =
+             e.target ∩ (e'.to_fun ∘ e.inv_fun ⁻¹' s ∩ e.inv_fun ⁻¹' e'.source),
+      by { rw [← inter_assoc, ← inter_assoc], congr' 1, exact inter_comm _ _ },
+    simpa [local_equiv.trans_source, preimage_inter, preimage_comp.symm, A] using this
+  end,
+  ..e }
+
+def to_manifold : @manifold H _ M c.to_topological_space :=
+{ atlas := ⋃ (e : local_equiv M H) (he : e ∈ c.atlas), {c.local_homeomorph e he},
+  chart_at := λx, c.local_homeomorph (c.chart_at x) (c.chart_mem_atlas x),
+  mem_chart_source := λx, c.mem_chart_source x,
+  chart_mem_atlas := λx, begin
+    simp only [mem_Union, mem_singleton_iff],
+    exact ⟨c.chart_at x, c.chart_mem_atlas x, rfl⟩,
+  end }
+
+end manifold_core
+
+section has_groupoid
+variables [topological_space H] [topological_space M] [manifold H M]
+
+/-- A manifold has an atlas in a groupoid G if the change of coordinates belong to the groupoid -/
+class has_groupoid (H : Type*) [topological_space H] (M : Type*) [topological_space M]
+  [manifold H M] (G : structure_groupoid H) : Prop :=
+(compatible : ∀e e' : local_homeomorph M H, e ∈ atlas H M → e' ∈ atlas H M → e.symm ≫ₕ e' ∈ G)
+
+lemma has_groupoid_of_le {G₁ G₂ : structure_groupoid H} (h : has_groupoid H M G₁) (hle : G₁ ≤ G₂) :
+  has_groupoid H M G₂ :=
+⟨ λ e e' he he', hle ((h.compatible : _) e e' he he') ⟩
+
+/-- The trivial manifold structure on the model space is compatible with any groupoid -/
+instance has_groupoid_model_space (H : Type*) [topological_space H] (G : structure_groupoid H) :
+  has_groupoid H H G :=
+{ compatible := λe e' he he', begin
+    replace he : e ∈ atlas H H := he,
+    replace he' : e' ∈ atlas H H := he',
+    rw model_space_atlas at he he',
+    simp [he, he', structure_groupoid.id_mem]
+  end }
+
+/-- Any manifold structure is compatible with the groupoid of all local homeomorphisms -/
+instance has_groupoid_continuous_groupoid : has_groupoid H M (continuous_groupoid H) :=
+⟨begin
+  assume e e' he he',
+  rw [continuous_groupoid, mem_groupoid_of_pregroupoid],
+  simp only [and_self]
+end⟩
+
+/-- In a manifold having some structure groupoid, the changes of coordinates belong to this groupoid -/
+lemma compatible (G : structure_groupoid H) [h : has_groupoid H M G]
+  {e e' : local_homeomorph M H} (he : e ∈ atlas H M) (he' : e' ∈ atlas H M) :
+  e.symm ≫ₕ e' ∈ G :=
+(h.compatible : _) e e' he he'
+
+/-- A G-diffeomorphism between two manifolds is a homeomorphism which, when read in the charts,
+belongs to G. -/
+structure diffeomorph (G : structure_groupoid H) (M : Type*) (M' : Type*)
```

Comment:
This name is *really* confusing. G could be the continuous groupoid for instance. Even if G is the symplectomorphism groupoid, or something like this, it's super confusing. We need to find a better name

Reply:
I agree it would be good to find a clearer name, but there is some (possibly bad) precedent --- I've written in a paper in G&T previously:
```
Finally, we need a general name for isomorphisms between balls, where the balls
could be piecewise linear or smooth or Spin or framed or etc., or some combination
thereof. We have chosen to use “homeomorphism” for the appropriate sort of 
isomorphism, so the reader should keep in mind that “homeomorphism” could 
mean PL homeomorphism or diffeomorphism (and so on) depending on context.
```

If we don't want to just announce that we are hereby generalising `homeomorph` or `diffeomorph`, I'll throw out: `structomorph`, `isomorph`, `local_structure_morphism`, `local_isomorph`, `G_morph`, `geomorphism`. These all seem either clunky or weird...

Reply:
I like `structomorph`. Let's got for this.

Reply:
`homeomorph` is already not really used in English, so `structomorph` cannot be that much worse :)

Reply:
Also, `structomorph` is a Latin-Greek hybrid compound, which is awesome (-;
https://en.wikipedia.org/wiki/Hybrid_word

Reply:
Let us never speak of this in the presence of my high school teacher, Mr Sheldon.